Build a regression model.

In [1]:
import pandas as pd 
import numpy as np
import json

Provide model output and an interpretation of the results. 

In [8]:
with open('df_final', 'r') as e: 
    data = json.load(e)
df_final = pd.DataFrame(data)
df_final

num_restaurant  num_entertainment  num_playground          lat  \
0                50                 47               1  51.52916347   
1                50                 18               1  51.49960695   
2                50                 50               1  51.52128377   
3                50                 50               2  51.52395143   
4                50                 50               2  51.52168078   
..              ...                ...             ...          ...   
540              50                  6               2  51.46663393   
541              50                 50               1    51.516204   
542              50                 50               1    51.517703   
543              50                 50               1  51.52364804   
544              50                 36               2  51.52536703   

             long                                           name  bikes  
0    -0.109970527            001023 - River Street , Clerkenwell     12  
1    -0.197574246        001018 - Phillimore Gardens, Kensington     31  
2    -0.084605692  001012 - Christopher Street, Liverpool Street      4  
3    -0.122502346          001022 - Brunswick Square, Bloomsbury     16  
4    -0.130431727              000980 - Malet Street, Bloomsbury      8  
..            ...                                            ...    ...  
540  -0.169821175         300014 - Falcon Road, Clapham Junction     28  
541     -0.155525            200008 - Portman Square, Marylebone      3  
542     -0.154106             001225 - George Street, Marylebone     25  
543  -0.074754872        000973 - Bethnal Green Road, Shoreditch     20  
544  -0.153463612  001199 - The Tennis Courts, The Regent's Park      1  

[545 rows x 7 columns]

In [4]:
df_final.describe()

num_restaurant    rating_x  num_starbucks    rating_y       bikes
count      797.000000  797.000000     797.000000  797.000000  797.000000
mean        41.089084    4.327478      26.974906    3.266625   13.858218
std         13.661590    0.361509      18.037980    0.821425    9.212250
min          3.000000    2.000000       1.000000    1.000000    0.000000
25%         32.000000    4.000000      10.000000    3.000000    7.000000
50%         50.000000    4.500000      23.000000    3.500000   13.000000
75%         50.000000    4.500000      50.000000    3.500000   19.000000
max         50.000000    5.000000      50.000000    5.000000   57.000000

In [9]:
import statsmodels.api as sm
X = df_final['num_restaurant']
y = df_final['bikes']
X = sm.add_constant(X)
model = sm.OLS(y, X)
results = model.fit()
print(results.summary())

                            OLS Regression Results                            
Dep. Variable:                  bikes   R-squared:                       0.002
Model:                            OLS   Adj. R-squared:                  0.000
Method:                 Least Squares   F-statistic:                     1.104
Date:                Wed, 08 Mar 2023   Prob (F-statistic):              0.294
Time:                        08:50:52   Log-Likelihood:                -1964.9
No. Observations:                 545   AIC:                             3934.
Df Residuals:                     543   BIC:                             3942.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                     coef    std err          t      P>|t|      [0.025      0.975]
----------------------------------------------------------------------------------
const             17.3521      4.092      4.

Initally I would assume that the number of restaurant would be a strong predicator of the number of bikes rented. However, statistically I was proven wrong. First, the R-square is 0.002 which means there is only a small portion of variability in "bikes rented" being variable beign explained by "number of restaurant" variable. This is also supported by the adjust R-squared value which is also very close to zero. Second, the P-value is 0.294 which is larger than the regular value of 0.05. this suggests that the linear relationship between "bikes rented" and "number of restaurant" is not significant. 

In [11]:
X = df_final['num_entertainment']
y = df_final['bikes']
X = sm.add_constant(X)
model = sm.OLS(y, X)
results = model.fit()
print(results.summary())

                            OLS Regression Results                            
Dep. Variable:                  bikes   R-squared:                       0.033
Model:                            OLS   Adj. R-squared:                  0.031
Method:                 Least Squares   F-statistic:                     18.52
Date:                Wed, 08 Mar 2023   Prob (F-statistic):           2.00e-05
Time:                        08:51:20   Log-Likelihood:                -1956.3
No. Observations:                 545   AIC:                             3917.
Df Residuals:                     543   BIC:                             3925.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                        coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------
const                16.0602      0.79

Out of all the three independent variables, the number of entertainment suggests reelatively stronger relationship with the number of bike rentals. Although the R-squared value is only 0.033, which means only 3.3% of the variation in the number of bikes rented can be explained by the number of entertaiment, but the result has a P-value of 2.00e -05. This indicates that there is a significant relationship between the number of entertainment and the number of bike rentals.  

In [12]:
X = df_final['num_playground']
y = df_final['bikes']
X = sm.add_constant(X)
model = sm.OLS(y, X)
results = model.fit()
print(results.summary())


                            OLS Regression Results                            
Dep. Variable:                  bikes   R-squared:                       0.005
Model:                            OLS   Adj. R-squared:                  0.003
Method:                 Least Squares   F-statistic:                     2.747
Date:                Wed, 08 Mar 2023   Prob (F-statistic):             0.0980
Time:                        09:40:59   Log-Likelihood:                -1964.1
No. Observations:                 545   AIC:                             3932.
Df Residuals:                     543   BIC:                             3941.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                     coef    std err          t      P>|t|      [0.025      0.975]
----------------------------------------------------------------------------------
const             14.3116      0.840     17.

The reason I chose playground as one of my independent variable is because I would assume that people usually take light transportation to nearby playgrounds. However, the result produced suggests otherwise. For example, the r-squared and adjuted r-squared values are extremely low, which means the nubmer of playgrounds is a weak predictor of the number of bike rentals. Furthermore, P-value is around 0.098 which also suggests that it is not statistically significant since it is greater than the 0.05 alpha value. In conlcusion, the model does not provide a good fit to the data. 

# Stretch

How can you turn the regression model into a classification model?